lib/_http_agent.js

```diff
@@ -112,11 +112,15 @@ function Agent(options) {
         if (this.sockets[name])
           count += this.sockets[name].length;
 
-        if (count > this.maxSockets || freeLen >= this.maxFreeSockets) {
+        if (count > this.maxSockets) {
           socket.destroy();
         } else if (this.keepSocketAlive(socket)) {
-          freeSockets = freeSockets || [];
-          this.freeSockets[name] = freeSockets;
+          if (freeLen >= this.maxFreeSockets) {
+            this.freeSockets[name].shift().destroy();
```

Comment:
nit, `freeSockets.shift().destroy()`

Also, what happens if `maxFreeSockets === 0`?

Reply:
Took another look at it and cleaned up the logic.